frontend/hooks/useRewardsHistory.ts

```diff
@@ -77,68 +82,80 @@ export type Checkpoint = {
-        reward = isRewardFinite ? currentReward ?? '0' : '0';
-      }
-
-      // If the epoch is 0, it means it's the first epoch else,
-      // the start time of the epoch is the end time of the previous epoch
-      const epochStartTimeStamp =
-        checkpoint.epoch === '0'
-          ? Number(checkpoint.blockTimestamp) - Number(checkpoint.epochLength)
-          : checkpoints[index + 1]?.blockTimestamp ?? 0;
-
-      const stakingContractId = getStakingProgramIdByAddress(
-        checkpoint.contractAddress as Address,
-      );
+const useTransformCheckpoints = () => {
+  const agent = useAgent();
+  const chainId = useChainId();
+
+  return useCallback(
+    (
+      serviceId: number,
+      checkpoints: CheckpointResponse[],
+      timestampToIgnore?: null | number,
+    ) => {
+      if (!checkpoints || checkpoints.length === 0) return [];
+      if (!serviceId) return [];
 
-      return {
-        ...checkpoint,
-        epochEndTimeStamp: Number(checkpoint.blockTimestamp ?? Date.now()),
-        epochStartTimeStamp: Number(epochStartTimeStamp),
-        reward: Number(ethers.utils.formatUnits(reward, 18)),
-        earned: serviceIdIndex !== -1,
-        contractName: stakingContractId,
-      };
-    })
-    .filter((checkpoint) => {
-      // If the contract has been switched to new contract,
-      // ignore the rewards from the old contract of the same epoch,
-      // as the rewards are already accounted in the new contract.
-      // Example: If contract was switched on September 1st, 2024,
-      // ignore the rewards before that date in the old contract.
-      if (!timestampToIgnore) return true;
-
-      if (!checkpoint) return false;
-      if (!checkpoint.epochEndTimeStamp) return false;
-
-      return checkpoint.epochEndTimeStamp < timestampToIgnore;
-    });
+      return checkpoints
+        .map((checkpoint: CheckpointResponse, index: number) => {
+          const serviceIdIndex =
+            checkpoint.serviceIds?.findIndex(
+              (id) => Number(id) === serviceId,
+            ) ?? -1;
+
+          let reward = '0';
+
+          if (serviceIdIndex !== -1) {
+            const currentReward = checkpoint.rewards?.[serviceIdIndex];
+            const isRewardFinite = isFinite(Number(currentReward));
+            reward = isRewardFinite ? currentReward ?? '0' : '0';
+          }
+
+          // If the epoch is 0, it means it's the first epoch else,
+          // the start time of the epoch is the end time of the previous epoch
+          const epochStartTimeStamp =
+            checkpoint.epoch === '0'
+              ? Number(checkpoint.blockTimestamp) -
+                Number(checkpoint.epochLength)
+              : checkpoints[index + 1]?.blockTimestamp ?? 0;
+
+          const stakingContractId =
+            agent.serviceApi.getStakingProgramIdByAddress(
+              chainId,
+              checkpoint.contractAddress as Address,
+            );
```

Comment:
there's no need to run this in the map for every checkpoint, will spam the rpc endpoint, can pull this once, as address never changes for a stakingProgram